src/Development/IDE/LSP/Outline.hs

```diff
@@ -0,0 +1,194 @@
+                       )
+    , _kind     = SkClass
+    , _detail   = Just "class"
+    , _children =
+      Just $ List
+        [ (defDocumentSymbol l :: DocumentSymbol)
+            { _name           = showRdrName n
+            , _kind           = SkMethod
+            , _selectionRange = srcSpanToRange l'
+            }
+        | L l  (ClassOpSig False names _) <- tcdSigs
+        , L l' n                            <- names
+        ]
+    }
+documentSymbolForDecl (L l (TyClD DataDecl { tcdLName = L _ name, tcdDataDefn = HsDataDefn { dd_cons } }))
+  = Just (defDocumentSymbol l :: DocumentSymbol)
+    { _name     = showRdrName name
+    , _kind     = SkStruct
+    , _children =
+      Just $ List
+        [ (defDocumentSymbol l :: DocumentSymbol)
+            { _name           = showRdrName n
+            , _kind           = SkConstructor
+            , _selectionRange = srcSpanToRange l'
+            }
+        | L l  x <- dd_cons
+        , L l' n <- getConNames x
+        ]
+    }
+documentSymbolForDecl (L l (TyClD SynDecl { tcdLName = L l' n })) = Just
+  (defDocumentSymbol l :: DocumentSymbol) { _name           = showRdrName n
+                                          , _kind           = SkTypeParameter
+                                          , _selectionRange = srcSpanToRange l'
+                                          }
+documentSymbolForDecl (L l (InstD (ClsInstD { cid_inst = ClsInstDecl { cid_poly_ty } })))
+  = Just (defDocumentSymbol l :: DocumentSymbol) { _name = pprText cid_poly_ty
+                                                 , _kind = SkInterface
+                                                 }
+documentSymbolForDecl (L l (InstD DataFamInstD { dfid_inst = DataFamInstDecl (HsIB { hsib_body = FamEqn { feqn_tycon, feqn_pats } }) }))
+  = Just (defDocumentSymbol l :: DocumentSymbol)
+    { _name = showRdrName (unLoc feqn_tycon) <> " " <> T.unwords
+                (map pprText feqn_pats)
+    , _kind = SkInterface
+    }
+documentSymbolForDecl (L l (InstD TyFamInstD { tfid_inst = TyFamInstDecl (HsIB { hsib_body = FamEqn { feqn_tycon, feqn_pats } }) }))
+  = Just (defDocumentSymbol l :: DocumentSymbol)
+    { _name = showRdrName (unLoc feqn_tycon) <> " " <> T.unwords
+                (map pprText feqn_pats)
+    , _kind = SkInterface
+    }
+documentSymbolForDecl (L l (DerivD DerivDecl { deriv_type })) =
+  gfindtype deriv_type <&> \(L (_ :: SrcSpan) name) ->
+    (defDocumentSymbol l :: DocumentSymbol) { _name = pprText @(HsType GhcPs)
+                                              name
+                                            , _kind = SkInterface
+                                            }
+documentSymbolForDecl (L l (ValD x)) =
+  gfindtype x <&> \(L (_ :: SrcSpan) name) ->
+    (defDocumentSymbol l :: DocumentSymbol)
+      { _name   = showRdrName name
+      , _kind   = case x of
+                    PatBind{} -> SkConstant
+                    _         -> SkFunction
```

Comment:
I'm not sure the distinction between `SkConstant` and `SkFunction` makes sense at this point. As I understand, a binding like `Just plus2 = Just (+ 2)` would be qualified as `SkConstant`, while `two = 2` would be qualified as `SkFunction`. Sadly, LSP doesn't seem to have `SkValue`, but it seems more to correct to consistently use either `SkFunction` or `SkConstant`.